### PR TITLE
History refactoring for TS

### PR DESCRIPTION
**Pull request by kewne7768**
This PR cleans up the history stuff to be more TypeScript-friendly. It also re-adds the full display of diamonds/mythos/particles instead of only showing the exponent (this was done because they didn't fit at some point in the past, but it's now hiding useful info for lategame players).

Submitted as draft because I want to fix the last remaining use of `any` and test it a bit more before having it merged.

### CODE REVIEW

File: src/History.ts [hunk 1]
```diff
@@ -1,51 +1,76 @@
...
-const formatDecimalString = (str: DecimalSource) => format(new Decimal(str));
+export type ResetHistoryEntryUnion =
+    ResetHistoryEntryAntSacrifice
+    | ResetHistoryEntryPrestige
+    | ResetHistoryEntryTranscend
+    | ResetHistoryEntryReincarnate
+    | ResetHistoryEntryAscend
+
+// A formatter that allows formatting a string. The string should be in a form parsable by break_infinity.js.
+const formatStringOrNumber = (numOrStr: number | Decimal | string) => {
```
Comment (KhafraDev) [inline src/History.ts:65]:
```suggestion
const formatStringOrNumber = (numOrStr: DecimalSource) => {
```

File: src/History.ts [hunk 2]
```diff
@@ -1,51 +1,76 @@
...
+    return format(typeof numOrStr === "string" ? new Decimal(numOrStr) : numOrStr);
+}
 
-const conditionalFormatPerSecond = (numOrStr: Decimal | number, data: ResetHistoryAscend | ResetHistoryDate) => {
+// A formatter that, if given a number, allows the data to be divided by the amount of seconds spent.
+const conditionalFormatPerSecond = (numOrStr: number | Decimal | string, data: ResetHistoryEntryBase) => {
```
Comment (KhafraDev) [inline src/History.ts:70]:
```suggestion
const conditionalFormatPerSecond = (numOrStr: DecimalSource, data: ResetHistoryEntryBase) => {
```

File: src/History.ts [hunk 3]
```diff
@@ -228,34 +235,38 @@ const resetHistoryRenderRow = (
     const gains = [];
     for (let gainIdx = 0; gainIdx < historyGainsOrder.length; ++gainIdx) {
         const showing = historyGainsOrder[gainIdx];
-        if (data.hasOwnProperty(showing)) {
+        // TODO: Rework this code so we don't have to any
+        const lolLmao = data as any;
```
Comment (KhafraDev) [inline src/History.ts:239]: this of course, as you mentioned ;)

File: src/Reset.ts [hunk 4]
```diff
@@ -163,17 +168,77 @@ export const updateTesseractAutoBuyAmount = () => {
...
-export const reset = (input: resetNames, fast = false, from = 'unknown') => {
-    const historyEntry: Partial<ResetHistoryAscend> = {};
-    let historyKind = "prestige";
-    const historyCategory = (input === 'ascension' || input === 'ascensionChallenge') ? 'ascend' : 'reset';
-    // By default, we don't log history entries when the player is entering or leaving a challenge, but we handle some
-    // special cases down below. This keeps the logs clean when someone in lategame runs 30 challenges in a row.
-    let historyUse = from !== "enterChallenge" && from !== "leaveChallenge";
+const resetAddHistoryEntry = (input: resetNames, from = 'unknown') => {
+    const offeringsGiven = calculateOfferings(input);
+    const isChallenge = ["enterChallenge", "leaveChallenge"].indexOf(from) !== -1;
```
Comment (KhafraDev) [inline src/Reset.ts:173]:
```suggestion
    const isChallenge = ["enterChallenge", "leaveChallenge"].includes(from);
```

File: src/Reset.ts [hunk 5]
```diff
@@ -225,10 +290,6 @@ export const reset = (input: resetNames, fast = false, from = 'unknown') => {
 
     if (input === "transcension" || input === "transcensionChallenge" || input == "reincarnation" || input == "reincarnationChallenge"
         || input === "ascension" || input === "ascensionChallenge") {
```
Comment (KhafraDev) [inline src/Reset.ts:292]:
```suggestion
    const types = ['transcension', 'transcensionChallenge', 'reincarnation', 'reincarnationChallenge', 'ascension', 'ascensionChallenge'];
    if (types.includes(input)) {
```

Comment (kewne7768) [reply src/Reset.ts:292]: I left this one alone because I didn't touch the code, but I might as well see if GitHub's "commit suggestion" button works

**Comment by kewne7768** [comment]
V2 of the patchset, fixing the use of `any` and just making it a lot stricter overall. I got a bit carried away once I started to clean things up more, so I changed more than I'd have liked. Also, I removed the `any` in the Player object's `.history` as well while I was at it.

I didn't update the files under `dist/`, not sure if it's a good idea to commit those during development because it seems like merge conflict hell.

**Review by KhafraDev — APPROVED** [review]

**Review by Pseudonian — APPROVED** [review]

**Comment by Pseudonian** [comment]
Thank you for doing the history refactoring for the TS rewrite. It works wonderfully and it'll help us expand upon the feature in the future. Merged!

**Pull request merged**